core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -230,13 +233,37 @@ public void clearLockedKeys() {
       lockedKeys = null;
    }
 
-   private boolean hasLockOrIsLockBackup(Object key) {
-      //stopgap fix for ISPN-2728. The real fix would be to synchronize this with the intrinsic lock.
+   @Override
+   public boolean containsLockOrBackupLock(Object key) {
       Set<Object> lockedKeysCopy = lockedKeys;
       Set<Object> backupKeyLocksCopy = backupKeyLocks;
       return (lockedKeysCopy != null && lockedKeysCopy.contains(key)) || (backupKeyLocksCopy != null && backupKeyLocksCopy.contains(key));
    }
 
+   @Override
+   public boolean containsAnyLockOrBackupLock(Stream<Object> keys) {
+      return findFirstLockedOrBackupLocked(keys) != null;
+   }
+
+   @Override
+   public Object findFirstLockedOrBackupLocked(Stream<Object> keys) {
+      Set<Object> lockedKeysCopy = lockedKeys;
+      Set<Object> backupKeyLocksCopy = backupKeyLocks;
+      if (lockedKeysCopy != null && backupKeyLocksCopy != null) {
+         return keys.filter(key -> lockedKeysCopy.contains(key) || backupKeyLocksCopy.contains(key)).findAny().orElse(null);
+      } else if (lockedKeysCopy != null) {
+         return keys.filter(lockedKeysCopy::contains).findAny().orElse(null);
+      } else if (backupKeyLocksCopy != null) {
+         return keys.filter(backupKeyLocksCopy::contains).findAny().orElse(null);
+      }
+      return null;
+   }
```

Comment:
>
> I can understand that yes. Also when using iteration, if using an array
> list we should use for int loops instead as they are also 16% faster.
>

I also did some more tests with forEach() and for int, and on my i7
forEach() seems to handle lists with a 5 elements a little better:

N=5
Benchmark Mode Cnt Score Error Units
StreamVsVanilla.forEach avgt 50 11.059 ± 0.144 ns/op
StreamVsVanilla.forInt avgt 50 13.175 ± 0.104 ns/op
StreamVsVanilla.vanilla avgt 50 12.583 ± 0.053 ns/op
N=1
Benchmark Mode Cnt Score Error Units
StreamVsVanilla.forEach avgt 50 32.876 ± 0.165 ns/op
StreamVsVanilla.forInt avgt 50 32.924 ± 0.521 ns/op
StreamVsVanilla.vanilla avgt 50 35.940 ± 0.207 ns/op

 Although we need to be cognizant of what the alternatives are more than
> just saying 1 thing is bad imo. For example iterating over the collection
> multiple times or creating a new collection could be worse then just using
> a stream.
>

Indeed, the alternative is not always cheaper. But for our needs, I think
it's easy enough to do the same stuff streams do without allocating new
objects. Like streams collect a bunch of operations in a pipeline
(allocating new objects in the process) and execute them all with a single
collection iteration, we can also collect operations in method parameters
(using only the stack) and execute them with a single collection iteration.